components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQMessageConverter.java

```diff
@@ -49,158 +52,196 @@
-
         Message msg;
         if (exchange.hasOut()) {
             msg = exchange.getOut();
         } else {
             msg = exchange.getIn();
         }
 
-        final Object contentType = msg.removeHeader(RabbitMQConstants.CONTENT_TYPE);
+        AMQP.BasicProperties.Builder properties = buildBasicAmqpProperties(exchange.getProperties(), msg);
+
+        final Map<String, Object> headers = msg.getHeaders();
+        // Add additional headers (if any)
+        if (additionalHeaders != null) {
+            headers.putAll(additionalHeaders);
+        }
+        Map<String, Object> filteredHeaders = new HashMap<>();
+
+        for (Map.Entry<String, Object> header : headers.entrySet()) {
+            // filter header values.
+            Object value = getValidRabbitMQHeaderValue(header.getKey(), header.getValue());
+
+            // additionally filter out the OVERRIDE header so it does not
+            // propagate
+            if ((value != null || isAllowNullHeaders()) && !header.getKey()
+                    .equals(RabbitMQConstants.EXCHANGE_OVERRIDE_NAME.key())) {
+                boolean filteredHeader;
+                if (!allowCustomHeaders) {
+                    filteredHeader = headerFilterStrategy
+                            .applyFilterToCamelHeaders(header.getKey(), header.getValue(), exchange);
+                    if (filteredHeader) {
+                        filteredHeaders.put(header.getKey(), header.getValue());
+                    }
+                } else {
+                    filteredHeaders.put(header.getKey(), header.getValue());
+                }
+            } else if (LOG.isDebugEnabled()) {
+                if (header.getValue() == null) {
+                    LOG.debug("Ignoring header: {} with null value", header.getKey());
+                } else if (header.getKey().equals(RabbitMQConstants.EXCHANGE_OVERRIDE_NAME.key())) {
+                    LOG.debug("Preventing header propagation: {} with value {}:", header.getKey(),
+                            header.getValue());
+                } else {
+                    LOG.debug("Ignoring header: {} of class: {} with value: {}", header.getKey(),
+                            ObjectHelper.classCanonicalName(header.getValue()), header.getValue());
+                }
+            }
+        }
+
+        properties.headers(filteredHeaders);
+
+        return properties;
+    }
+
+    private AMQP.BasicProperties.Builder buildBasicAmqpProperties(
+            Map<String, Object> exchangeProperties, Message msg) {
+        AMQP.BasicProperties.Builder properties = new AMQP.BasicProperties.Builder();
+
+        Object contentType = getBasicAmqpProperty(exchangeProperties, msg, RabbitMQConstants.CONTENT_TYPE);
         if (contentType != null) {
             properties.contentType(contentType.toString());
         }
 
-        final Object priority = msg.removeHeader(RabbitMQConstants.PRIORITY);
```

Comment:
Any reason why this should be mutable by removing `final`?

Reply:
Removed it accidentally. It should be there. Accepted.